base/operators.jl

```diff
@@ -985,6 +985,54 @@ function show(io::IO, c::ComposedFunction)
     show(io, c.inner)
 end
 
+"""
+    IteratedFunction{F} <: Function
+
+`IteratedFunction(f,n)` represents the function `f` iterated `n ≥ 0` times on
+its input, which must be a single argument.
+
+That is, for functions `f(x)`, it represents the function
+`x -> f(f(f(f(...(f(x))))))` iterated `n` times`.  Any keyword arguments
+are passed through to all calls.
+
+If `f isa Function`, you should normally use the construction `f^n`
+to form an `IteratedFunction`.  If `n` is a literal integer, `f^n`
+may construct a more specialized object, e.g. `f^1 == f`, `f^0 == identity`,
+and `f^2 == f ∘ f`.
+
+!!! compat "Julia 1.7"
+    `IteratedFunction` and `f^n` require at least Julia 1.7.
+
+See also [`∘`](@ref) and [`ComposedFunction`](@ref).
+"""
+struct IteratedFunction{F} <: Function
+    f::F
+    n::Int
```

Comment:
Maybe this should be a type parameter instead, so application can be type stable?

Reply:
I thought about that — it depends on how this is used.   If `n` is almost always a hard-coded literal value, then a type parameter is definitely the way to go.  On the other hand, if people use an `n` determined dynamically a lot, it would create a lot of dynamic dispatch.

Reply:
Maybe one could get the best of both words?   Use `n::TN` where `TN` is `Val{N}` if a literal argument was used, and where `TN` is `Int` otherwise?

Reply:
Is it really that common to construct a lot of these with dynamic `n`? The uses I had in mind was mostly with higher order functions like `map`, where the cost of calling the functor is much more important than that of constructing and dynamically passing it. Making this dependent on `literal_pow` sounds interesting, but I can still imagine use cases, where `n` is not a literal, but will still pretty much always constant propagate. I am also not 100% sure whether we want to make `literal_pow` more special, since it's not an uncommon source of confusion already.